### PR TITLE
Implement support for OGC API Features

### PR DESCRIPTION
**Pull request by SpeckiJ**
depends on:

 - https://github.com/KomMonitor/api-specs/commits/feature/openapi-3.0.x
 -

### CODE REVIEW

**Review by SebaDro — CHANGES_REQUESTED** [review]

File: kommonitor-models/src/main/java/org/n52/kommonitor/models/ResponseEntity.java
[the remark this answers is not in the document]
Comment (SpeckiJ) [reply kommonitor-models/src/main/java/org/n52/kommonitor/models/ResponseEntity.java:1]: Is there a way to exclude this file from auto-generation? I don't like the workflow of manually deleting the file everytime we regenerate the api-specs.

If this is not trivially possible i would just leave the file there. During development you will very quickly notice that you imported the wrong `ResponseEntity` as it does not provide the same methods and is a totally different type of class.

File: kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+            ArrayList<FeatureCollection.Feature> features = new ArrayList<>();
+
+            String bbox;
+            Geometry referenceGeometry;
+
+            // Get spatialUnit by reference
+            if (bboxType.equals("ref")) {
+
+                // Get spatialUnit from data-management
+                String response = apiClient.getAllSpatialUnitFeaturesById(bboxValue, null, "strong");
```
[the remark this answers is not in the document]
Comment (SpeckiJ) [reply kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java]: fixed in https://github.com/KomMonitor/kommonitor-importer/pull/25/commits/87a9ce180d874a1e7b07415e30b86f5d6bfbc4d8

File: kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/converter/APIFeaturesConverter.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    public Set<String> initSupportedDatasources() {
+        Set<String> types = new HashSet<>();
+        types.add(DataSourceType.TypeEnum.OGCAPI_FEATURES.getValue());
+        return types;
+    }
+
+    @Override
+    public Set<ConverterParameter> initConverterParameters() {
+        return Collections.emptySet();
+    }
```
[the remark this answers is not in the document]
Comment (SpeckiJ) [reply kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/converter/APIFeaturesConverter.java]: fixed in https://github.com/KomMonitor/kommonitor-importer/pull/25/commits/c5df4d04a9aed018ef1a2e36a444ae3b854b54a6 . We now inherit the parameters from `GeoJsonConverter`

File: kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+    private static final String TYPE = DataSourceType.TypeEnum.OGCAPI_FEATURES.getValue();
+    private static final String PARAM_URL = "url";
+    private static final String PARAM_BBOX = "bbox";
+    private static final String PARAM_BBOX_TYPE = "bboxType";
+    private static final String PARAM_CUSTOM_FILTER = "filter";
+    private static final String PARAM_URL_DESC = "An URL that references a OGC API - Features dataset. " +
+            "The dataset will be retrieved with HTTP GET requests for that URL.";
+    private static final String PARAM_BBOX_DESC = "A bounding box for this dataset. Either a literal array of four" +
+            " coordinates (bboxType=='ref') , or uuid of spatialUnit used as reference (bboxType=='literal')";
```
[the remark this answers is not in the document]
Comment (SpeckiJ) [reply kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java]: improved to documentation in https://github.com/KomMonitor/kommonitor-importer/pull/25/commits/d845002083372a3aae8bf0c8c8ea2dbe7af4ecb6

File: kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    @Override
+    protected String initType() {
+        return TYPE;
+    }
+
+    @Override
+    protected Set<DataSourceParameter> initSupportedParameters() {
+        Set<DataSourceParameter> parameters = new HashSet<>();
+        parameters.add(new DataSourceParameter(PARAM_BBOX, PARAM_BBOX_DESC, DataSourceParameter.ParameterTypeValues.ARRAY));
+        parameters.add(new DataSourceParameter(PARAM_BBOX_TYPE, PARAM_BBOX_DESC, DataSourceParameter.ParameterTypeValues.ARRAY));
```
[the remark this answers is not in the document]
Comment (SpeckiJ) [reply kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java]: fixed in https://github.com/KomMonitor/kommonitor-importer/pull/25/commits/d845002083372a3aae8bf0c8c8ea2dbe7af4ecb6

File: kommonitor-models/src/main/java/org/n52/kommonitor/models/ResponseEntity.java
[the remark this answers is not in the document]
Comment (SebaDro) [reply kommonitor-models/src/main/java/org/n52/kommonitor/models/ResponseEntity.java:1]: Yes. Just add the class name to the `.openapi-generator-ignore` file: https://github.com/OpenAPITools/openapi-generator/blob/master/docs/customization.md#ignore-file-format

File: kommonitor-importer-core/src/test/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetrieverTest.java
[the remark this answers is not in the document]
Comment (SebaDro) [reply kommonitor-importer-core/src/test/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetrieverTest.java:1]: Besides that, I noticed that the Unit Test fails which leads to a failing build.

File: kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+            // Get spatialUnit by reference
+            if (bboxType.equals("ref")) {
+
+                // Get spatialUnit from data-management
+                String response = apiClient.getAllSpatialUnitFeaturesById(bboxValue, null, "original");
+
+                // Extract bbox for first stage filtering (intersection with bbox)
+                DefaultFeatureCollection featureCollection = (DefaultFeatureCollection) featureJSON.readFeatureCollection(response);
+                ReferencedEnvelope envelope = featureCollection.getBounds();
+                bbox = String.format("%f,%f,%f,%f", envelope.getMinX(), envelope.getMinY(), envelope.getMaxX(), envelope.getMaxY());
```
Comment (SebaDro) [inline kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java:139]: This string formatting results in a format such as `6,067986,51,166833,6,625991,51,425020` with comma as decimal separator. However, OGC API - Feature services expects a decimal point. So a generated requests, such as https://ogc-api.nrw.de/brgnrw/v1/collections/Bohrungen/items?f=json&bbox=6%2C067986%2C51%2C166833%2C6%2C625991%2C51%2C425020 results in an error response:
```json
{
"status": 400,
"title": "Bad Request",
"detail": "The following query parameters are rejected: bbox. Valid parameters for this request are: access_token, f"
}
```

**Review by SebaDro — CHANGES_REQUESTED** [review]
There seems to be a problem qith formatting the bbox query parameter. Can you have a look on it?

File: kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java (hunk 5 above)
Comment (SpeckiJ) [reply kommonitor-importer-core/src/main/java/org/n52/kommonitor/importer/io/datasource/APIFeaturesRetriever.java]: fixed in https://github.com/KomMonitor/kommonitor-importer/pull/25/commits/19cfbec0b74ed9fac7d64b0f0b442ea0e366819a

**Review by SebaDro — APPROVED** [review]

**Pull request merged**